les/catalyst/api.go

```diff
@@ -66,124 +77,217 @@ func NewConsensusAPI(les *les.LightEthereum) *ConsensusAPI {
+		finalHeader := api.les.BlockChain().GetHeaderByHash(update.FinalizedBlockHash)
+		if finalHeader == nil {
+			log.Warn("Final block not available in database", "hash", update.FinalizedBlockHash)
+			return beacon.STATUS_INVALID, beacon.InvalidForkChoiceState.With(errors.New("final block not available in database"))
+		} else if rawdb.ReadCanonicalHash(api.les.ApiBackend.ChainDb(), finalHeader.Number.Uint64()) != update.FinalizedBlockHash {
+			log.Warn("Final block not in canonical chain", "number", header.Number.Uint64(), "hash", update.HeadBlockHash)
+			return beacon.STATUS_INVALID, beacon.InvalidForkChoiceState.With(errors.New("final block not in canonical chain"))
+		}
+		// Set the finalized block
+		// TODO (MariusVanDerWijden): Enable this once Finalized is implemented in LES
+		// api.les.BlockChain().SetFinalized(finalBlock)
+	}
+	// Check if the safe block hash is in our canonical tree, if not somethings wrong
+	if update.SafeBlockHash != (common.Hash{}) {
+		safeHeader := api.les.BlockChain().GetHeaderByHash(update.SafeBlockHash)
+		if safeHeader == nil {
+			log.Warn("Safe block not available in database")
+			return beacon.STATUS_INVALID, beacon.InvalidForkChoiceState.With(errors.New("safe block not available in database"))
+		}
+		if rawdb.ReadCanonicalHash(api.les.ApiBackend.ChainDb(), safeHeader.Number.Uint64()) != update.SafeBlockHash {
+			log.Warn("Safe block not in canonical chain")
+			return beacon.STATUS_INVALID, beacon.InvalidForkChoiceState.With(errors.New("safe block not in canonical chain"))
+		}
+	}
+	// Payload generation is not supported in light mode.
 	if payloadAttributes != nil {
+		log.Error("Block production requested in light mode")
 		return beacon.STATUS_INVALID, errors.New("not supported")
 	}
-	return api.validForkChoiceResponse(), nil
+	return valid(nil), nil
+}
+
+// ExchangeTransitionConfigurationV1 checks the given configuration against
+// the configuration of the node.
+func (api *ConsensusAPI) ExchangeTransitionConfigurationV1(config beacon.TransitionConfigurationV1) (*beacon.TransitionConfigurationV1, error) {
+	if config.TerminalTotalDifficulty == nil {
+		return nil, errors.New("invalid terminal total difficulty")
+	}
+	ttd := api.les.BlockChain().Config().TerminalTotalDifficulty
+	if ttd.Cmp(config.TerminalTotalDifficulty.ToInt()) != 0 {
+		log.Warn("Invalid TTD configured", "geth", ttd, "beacon", config.TerminalTotalDifficulty)
+		return nil, fmt.Errorf("invalid ttd: execution %v consensus %v", ttd, config.TerminalTotalDifficulty)
+	}
+
+	if config.TerminalBlockHash != (common.Hash{}) {
+		if hash := api.les.BlockChain().GetCanonicalHash(uint64(config.TerminalBlockNumber)); hash == config.TerminalBlockHash {
+			return &beacon.TransitionConfigurationV1{
+				TerminalTotalDifficulty: (*hexutil.Big)(ttd),
+				TerminalBlockHash:       config.TerminalBlockHash,
+				TerminalBlockNumber:     config.TerminalBlockNumber,
+			}, nil
+		}
```

Comment:
What if the terminal header is not available locally? It can happen since light client syncs from checkpoint.